File: Array/HIndexCalculator/cpp/HIndexCalculator.cpp
// Problem: https://leetcode.com/problems/h-index/

#include<vector>

using namespace std;

class HIndexCalculator {
public:
	int hIndex(vector<int>& citations) {
		int h_index = 0;
		int c_count[1001];
		for (int i = 0; i < 1001; ++i) {
			c_count[i] = 0;
		}
		for (int i = 0; i < citations.size(); ++i) {
			c_count[citations[i]]++;
		}
		int c_sum = 0;
		for (int i = 1000; i > 0; --i) {
			c_sum += c_count[i];
			if (i <= c_sum) { h_index = i; break; }
		}
		return h_index;
	}
};

File: Array/HIndexCalculator/go/HIndexCalculator.go
package _go

import (
	"sort"
)

// hIndex store the index for all of citation, if greater than length inc else inc the value one
// Time complexity: O(n)
// Space complexity: O(n)
func hIndex(citations []int) int {
	bucket := make(map[int]int, len(citations)+1)
	for _, v := range citations {
		if v > len(citations) {
			bucket[len(citations)]++
		} else {
			bucket[v]++
		}
	}
	count := 0
	for i := len(citations); i >= 0; i-- {
		count += bucket[i]
		if count >= i {
			return i
		}
	}
	return 0
}

// hIndex2 sort the citation and binary search
// Time complexity: O(nlog(n))
// Space complexity: O(1)
func hIndex2(citations []int) int {
	low, high := 0, len(citations)-1
	sort.Ints(citations)
	for low <= high {
		mid := low + (high-low)/2
		if citations[mid] >= len(citations)-mid {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}
	return len(citations) - low
}
